File: Golang/448. Find All Numbers Disappeared in an Array.go
// func findDisappearedNumbers(nums []int) []int { // 868ms???
//     for _, num := range nums {
//         index := int(math.Abs(float64(num)) - 1)
//         nums[index] = int(-math.Abs(float64(nums[index])))
//     }
//     ans := make([]int, 0, len(nums))
//     for i, num := range nums {
//         if num > 0 {
//             ans = append(ans, i+1)
//         }
//     }
//     return ans
// }

func findDisappearedNumbers(nums []int) []int { // 852ms ???换个时间执行一下
    for i := range nums {
        for nums[i] != nums[nums[i]-1] {
            nums[i], nums[nums[i]-1] = nums[nums[i]-1], nums[i]
        }
    }
    
    ans := make([]int, 0, len(nums))
    for i, num := range nums {
        if num != i+1 {
            ans = append(ans, i+1)
        }
    }
    return ans
}

File: Python3/448. Find All Numbers Disappeared in an Array.py
class Solution:
    def findDisappearedNumbers(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        # for num in nums: # 220ms
        #     index = abs(num) - 1
        #     nums[index] = -abs(nums[index])
        # return [i + 1 for i in range(len(nums)) if nums[i] > 0]
        
        for i in range(len(nums)): # 252ms
            while nums[i] != nums[nums[i]-1]:
                # nums[i], nums[nums[i]-1] = nums[nums[i]-1], nums[i] # 是错误的
                temp = nums[i]
                nums[i] = nums[temp-1]
                nums[temp-1] = temp      
        ans = []
        for i in range(len(nums)):
            if nums[i] != i+1:
                ans.append(i+1)
        return ans
